File: P2/python/algorithm.py
# Imports

import csv
import numpy as np
import copy

# Functions

def fun(x, beta):
	if x < 0:
		return 1

	return np.exp(-beta * x)

def dist(a, b):
	return (sum([(a[i] - b[i])**2 for i in range(0, len(a))]))**(1 / 2)

def length(path):
	l = float(0)

	for i in range(0, len(path)):
		l += dist(path[i - 1], path[i])

	return l

def var_length(path, n, i, j):
	var = 0

	if (j == n - 1) and (i == 0):
		i, j = j, i

	var -= dist(path[i - 1], path[i])
	var -= dist(path[j], path[(j + 1) % n])

	var += dist(path[i - 1], path[j])
	var += dist(path[i], path[(j + 1) % n])

	if j != (i + 1) % n:
		var -= dist(path[i], path[i + 1])
		var -= dist(path[j - 1], path[j])

		var += dist(path[j], path[i + 1])
		var += dist(path[j - 1], path[i])

	return var

# Read

x = list()

with open('resources/csv/x_best.csv', 'r') as f:
	r = csv.reader(f)
	for point in r:
		x.append((float(point[0]), float(point[1])))

# Metropolis-Hastings

## Parameters

m = 1e6

beta = 10
d_beta = 50 / m

## Initialization

n = len(x)
l = length(x)

x_min = copy.copy(x)
l_min = l
l_best = l

## Computations

print("Current best : %.6f km" %l_best)

while m > 0:
	m -= 1
	beta += d_beta

	i = np.random.randint(n - 1)
	j = np.random.randint(i + 1, n)

	delta_l = var_length(x, n, i, j)

	alpha = fun(delta_l, beta)

	if (alpha >= 1) or (np.random.rand() < alpha):
		x[i], x[j] = x[j], x[i]
		l += delta_l

		if l < l_min:
			x_min = copy.copy(x)
			l_min = l

# Write

if l_min < l_best:
	print("New best : %.6f km" %l_min)
	with open('resources/csv/x_best.csv', 'w', newline='\n') as f:
		w = csv.writer(f)
		for point in x:
			w.writerow([point[0], point[1]])

File: P2/matlab/algorithm.m
%% Load

name = 'belgium';
file = ['products/mat/' name '.mat'];

if exist(file, 'file') == 2
    load(file);
else
    initialize;
end

%% Metropolis-Hastings

% Initialization

[x, x_min, x_i] = deal(x_nna);
[l, l_min, l_i] = deal(l_nna);

% Parameters

m = 1e6;
c = min(1e6, m);
beta = 10;
beta_rel = beta * n / l;

% Algorithm

k = 1;

tic
while k < m
    if mod(k - 1, c) == 0
        I = randi([1 n], [c 2]);
    end
    
    i = I(mod(k - 1, c) + 1, 1);
    j = I(mod(k - 1, c) + 1, 2);
        
    delta = delta_f(n, D, x, i, j);
    
    a = alpha(beta_rel, delta);
    
    if a >= 1 || rand() < a
        l = l + delta;
        temp = x(i);
        x(i) = x(j);
        x(j) = temp;
        
        if l < l_min
            l_min = l;
            x_min = x;
        end
    end
    
    k = k + 1;
end
toc

%% Display

disp(['Initial length : ' num2str(l_i)]);
disp(['Nearest neighbour algorithm length : ' num2str(l_nna)]);
disp(['Best length in ' num2str(m) ' iterations : ' num2str(l_min)]);

% plot([tsp{x_min, 2}; tsp{x_min(1),2}], -[tsp{x_min, 3}; tsp{x_min(1),3}], '-');

%% Functions

function a = alpha(beta, x)
    if x < 0
        a = 1;
    else
        a = exp(-beta * x);
    end
end

function delta = delta_f(n, D, x, i, j)
    if i == j
        delta = 0; 
        return;
    elseif i > j
        temp = i;
        i = j;
        j = temp;
    end
    
    if i == 1 && j == n
        i_m = i + 1;
        j_p = j - 1;
    elseif i == 1
        i_m = n;
        j_p = j + 1;
    elseif j == n
        i_m = i - 1;
        j_p = 1;
    else
        i_m = i - 1;
        j_p = j + 1;
    end
    
    delta = -D(x(i_m), x(i));
    delta = delta - D(x(j), x(j_p));
    
    delta = delta + D(x(i_m), x(j));
    delta = delta + D(x(i), x(j_p));
    
    if i ~= j - 1 && (i ~= 1 || j ~= n)
        delta = delta - D(x(i), x(i + 1));
        delta = delta - D(x(j - 1), x(j));

        delta = delta + D(x(j), x(i + 1));
        delta = delta + D(x(j - 1), x(i));
    end
end
